File: code/algo/getoffer/c++/2/arr_collision.cpp
#include <vector>

std::vector<int> asteroidCollision(const std::vector<int> &asteroids) {
  std::vector<int> st;
  for (int aster : asteroids) {
    bool alive = true;
    while (alive && aster < 0 && !st.empty() && st.back() > 0) {
      alive = st.back() < -aster; // Check if the current asteroid survives
      if (st.back() <= -aster) {  // Top of the stack asteroid explodes
        st.pop_back();
      }
    }
    if (alive) {
      st.push_back(aster);
    }
  }
  return st;
}

File: code/algo/getoffer/go/2/arr_collision.go
package main

func asteroidCollision(asteroids []int) (st []int) {
	for _, aster := range asteroids {
		alive := true
		for alive && aster < 0 && len(st) > 0 && st[len(st)-1] > 0 {
			alive = st[len(st)-1] < -aster // aster 是否存在
			if st[len(st)-1] <= -aster {   // 栈顶小行星爆炸
				st = st[:len(st)-1]
			}
		}
		if alive {
			st = append(st, aster)
		}
	}
	return
}
